xray/segment.go

```diff
@@ -164,15 +158,44 @@ func (seg *Segment) RemoveSubsegment(remove *Segment) bool {
 	return false
 }
 
+func (seg *Segment) handleContextDone() {
+	seg.Lock()
+	defer seg.Unlock()
+
+	seg.ContextDone = true
+	if !seg.InProgress && !seg.Emitted {
+		seg.flush()
+	}
+}
+
+func (seg *Segment) flush() {
+	if (seg.openSegments == 0 && seg.EndTime > 0) || seg.ContextDone {
+		if seg.parent == nil {
+			seg.Emitted = true
+			emit(seg)
```

Comment:
was it important before that the segment was unlocked before `emit(seg)`?

Reply:
yeah because the emitter would lock the segment before chunking it into pieces and sending it off.